From 2d976bbd504f2fee21dea3f585e7c8951b41ea9e Mon Sep 17 00:00:00 2001
From: Akshay Pradip Bharad <apbharad@habana.ai>
Date: Fri, 23 May 2025 10:11:48 +0000
Subject: [PATCH] [SW-227647] ipm: recovery mechanism for spurious intr

This patch adds folllowing recovery mechanism:
1. Adds 3 rechecks after getting spurious intr.
2. If rechecks also fail, ignore the sequence number
and processes the msg irrspectively.

Change-Id: I9635bbfe2e47cf3a4e822b85c1ad3ea6f50ba2aa
Signed-off-by: Akshay Pradip Bharad <apbharad@habana.ai>
---

diff --git a/drivers/ipm/ipm_hl.c b/drivers/ipm/ipm_hl.c
index 5684d7c..e7697fa 100644
--- a/drivers/ipm/ipm_hl.c
+++ b/drivers/ipm/ipm_hl.c
@@ -17,6 +17,8 @@
 #define LOG_MODULE_NAME			ipm_hl
 LOG_MODULE_REGISTER(LOG_MODULE_NAME, CONFIG_IPM_LOG_LEVEL);
 
+#define IPM_ISR_FETCH_RETRY_COUNT 3
+
 /*
  * Beware that this function is called from ISR. It processes the message and
  * sends an ACK to the sender.
@@ -45,7 +47,7 @@
  * and process it. If the message is not found, it returns false.
  */
 
-static bool find_and_process_message(const struct device *dev)
+static int find_and_process_message(const struct device *dev, bool ignore_seq_no)
 {
 	const struct ipm_hl_config *config = DEV_CFG(dev);
 	struct ipm_hl_data *data = DEV_DATA(dev);
@@ -57,9 +59,9 @@
 			   (i * data->chunk_hdr_sz));
 
 		if ((msg->busy) && (!msg->ack)) {
-			if (msg->seq_no == data->rx_ctrl->ci) {
+			if (msg->seq_no == data->rx_ctrl->ci || ignore_seq_no) {
 				process_message(dev, msg);
-				return true;
+				return 0;
 			} else {
 				LOG_DBG("Out of Order Msg. seq_no: %d expected: %d",
 					msg->seq_no, data->rx_ctrl->ci);
@@ -72,9 +74,9 @@
 						(i * config->rx.data_chunk_size));
 
 		if ((msg->busy) && (!msg->ack)) {
-			if (msg->seq_no == data->rx_ctrl->ci) {
+			if (msg->seq_no == data->rx_ctrl->ci || ignore_seq_no) {
 				process_message(dev, msg);
-				return true;
+				return 0;
 			} else {
 				LOG_DBG("Out of Order Msg. seq_no: %d expected: %d",
 					msg->seq_no, data->rx_ctrl->ci);
@@ -82,23 +84,42 @@
 		}
 	}
 
-	LOG_WRN("Spurious interrupt received. PI: %d, CI: %d",
-		data->rx_ctrl->pi, data->rx_ctrl->ci);
-	return false;
+	if (!ignore_seq_no) {
+		LOG_WRN("Spurious interrupt received. PI: %d, CI: %d",
+			data->rx_ctrl->pi, data->rx_ctrl->ci);
+	}
+
+	return -ENOMSG;
 }
 
 void ipm_hl_isr(const struct device *dev)
 {
 	struct ipm_hl_data *data = DEV_DATA(dev);
-	uint32_t num_of_msgs = data->rx_ctrl->pi - data->rx_ctrl->ci;
+	uint32_t num_of_msgs = 0;
+	bool err, ignore_seq_no = false;
+
+	uint8_t retry = IPM_ISR_FETCH_RETRY_COUNT;
+
+	/* Try a few times to catch up in case of race with producer */
+	do {
+		num_of_msgs = data->rx_ctrl->pi - data->rx_ctrl->ci;
+		if (num_of_msgs > 0) {
+			break;
+		}
+		k_busy_wait(1);
+	} while (retry--);
 
 	if (num_of_msgs == 0) {
 		LOG_WRN("Spurious interrupt received");
-		return;
 	}
 
-	while (num_of_msgs > 0) {
-		if (!find_and_process_message(dev)) {
+	if (num_of_msgs != 1) {
+		ignore_seq_no = true;
+	}
+
+	while (num_of_msgs > 0 || ignore_seq_no) {
+		err = find_and_process_message(dev, ignore_seq_no);
+		if (err || ignore_seq_no) {
 			break;
 		}
 		num_of_msgs = data->rx_ctrl->pi - data->rx_ctrl->ci;
